### DISCUSSION TITLE
Upgrading from v10 to v12: ValidationError: Invalid configuration object.

### DISCUSSION THREAD

**Discussion by cschratz** [vercel/next.js#37590] [Help]
Working on upgrading my Next project from v10 to v12. Followed the instructions per the website on upgrading from 11 to 12 and I'm currently getting the following error when running 'npm run dev'. 

```
ValidationError: Invalid configuration object. Webpack has been initialized using a configuration object that does not match the API schema.
 - configuration[0].module.rules[6] should be one of these:
   ["..." | object { assert?, compiler?, dependency?, descriptionData?, enforce?, exclude?, generator?, include?, issuer?, issuerLayer?, layer?, loader?, mimetype?, oneOf?, options?, parser?, realResource?, resolve?, resource?, resourceFragment?, resourceQuery?, rules?, scheme?, sideEffects?, test?, type?, use? }, ...]
   -> A rule.
   Details:
    * configuration[0].module.rules[6].loader should be a non-empty string.
      -> A loader request.
 - configuration[1].module.rules[5] should be one of these:
   ["..." | object { assert?, compiler?, dependency?, descriptionData?, enforce?, exclude?, generator?, include?, issuer?, issuerLayer?, layer?, loader?, mimetype?, oneOf?, options?, parser?, realResource?, resolve?, resource?, resourceFragment?, resourceQuery?, rules?, scheme?, sideEffects?, test?, type?, use? }, ...]
   -> A rule.
   Details:
    * configuration[1].module.rules[5].loader should be a non-empty string.
      -> A loader request.
 - configuration[2].module.rules[6] should be one of these:
   ["..." | object { assert?, compiler?, dependency?, descriptionData?, enforce?, exclude?, generator?, include?, issuer?, issuerLayer?, layer?, loader?, mimetype?, oneOf?, options?, parser?, realResource?, resolve?, resource?, resourceFragment?, resourceQuery?, rules?, scheme?, sideEffects?, test?, type?, use? }, ...]
   -> A rule.
   Details:
    * configuration[2].module.rules[6].loader should be a non-empty string.
      -> A loader request.
```

The only changes I have made besides upgrading Next was to also upgrade React to 18.1 and Typescript to 4.3.2

Any help would be appreciated, thanks!

**Comment by icyJoseph** [comment]
I think somewhere in between the Webpack version upgrades to 5, and that has of course breaking changes and  a bunch of shapes have been updated. 

It is a bit hard to write an upgrade guide that takes into account every possible custom webpack config.

Would you mind sharing the webpack key of the next.config file? If you don't have any, but use plugins, perhaps you gotta update those. React and TS upgrades are not to blame here, it is all about that Webpack config!

**Reply by cschratz** [reply]
Yeah I was suspecting it was Webpack just based on the error. Here's the next.config file. No mention of webpack... 

```
module.exports = withWPEHeadless({
	trailingSlash: true,
	assetPrefix: assetPrefix,
	reactStrictMode: true,
});
```

withWPEHeadless looks like so:

```
module.exports = function withWPEHeadless(nextConfig = {}) {
  const withTM = require('next-transpile-modules')(['@wpengine/headless']);
  return withTM(nextConfig);
}
```

Could it possibly be due to the next-transpile-modules?

**Reply by icyJoseph** [reply]
Most likely! Over at their github page they have a table that tells you what version to use for each Next.js version, for Next 11 you need version 8.x.x., github.com/martpie/next-transpile-modules#readme

**Reply by icyJoseph** [reply]
And for 12, anything above 9.

**Reply by cschratz** [reply]
Yeah looks like the withWPEHeadless function from an external package (@wpengine/headless) is only using 6.0.0. Seems like the current project design won't support migration to 12.

**Reply by cschratz** [reply]
Updates:

So I was able to make the next-transpile-modules work, but when I try to run next dev it responds with:

```
‘Uncaught ReferenceError: exports is not defined’ 
```

Rolling back to Next 11 results in no issues, what changes could cause this to happen in Next 12 vs 11?

**Reply by icyJoseph** [reply]
ES modules support might be getting in the way... 

I copied your code, and installed `@wpengine/headless` and `next-transpile-modules` and everything works fine. Is there anything I have to do to get it crash? Perhaps import something from `@wpengine/headless`.

On a side note, `@wpengine/headless` has been deprecated for a while now, so I wouldn't put too much hope onto this process. This is why I don't know how to go forward, the documentation for this plugin is gone, and it has about 4 downloads a week..

> WordPress Headless Framework
> The [@wpengine/headless](https://www.npmjs.com/package/@wpengine/headless) package is deprecated. Please use [@faustjs/core](https://www.npmjs.com/package/@faustjs/core), [@faustjs/react](https://www.npmjs.com/package/@faustjs/react), and [@faustjs/next](https://www.npmjs.com/package/@faustjs/next) instead.
>
> Please see https://github.com/wpengine/faustjs for more information.

https://github.com/wpengine/faustjs/tree/main/packages/headless

**Reply by cschratz** [reply]
So in the __app.tsx file, there's an import for the Headless Provider component from @wpengine and then wrap the entire JSX return in that <HeadlessProvider> component. I'm thinking that might be the issue. 

Error on localhost

<img width="996" alt="Screen Shot 2022-06-10 at 8 06 50 AM" src="https://user-images.githubusercontent.com/60720268/173071041-48386a4c-c598-4b0d-b4ce-2893fee585f2.png">

First line of @wpengine/headless/dist/api/services.js

<img width="1398" alt="Screen Shot 2022-06-10 at 8 07 11 AM" src="https://user-images.githubusercontent.com/60720268/173071068-30ff7ac3-2d59-4b06-ad65-a1c5a63502d3.png">

**Reply by icyJoseph** [reply]
Aha, but then I think the solution is to stop using `next-transpile-modules` for the `@wpengine/headless` package.

**Reply by cschratz** [reply]
Okay found the solution. 

I had to add a .babelrc file in order to force Next to use the Babel compiler over the new SWC based compiler. 

After that I used the overrides property in package.json to force @wpegine/headless to utilize next-transpile-modules at v9.0.0

Maybe this information will help someone else down the line!
